docs/en/guides/90-community/02-rfcs/20250109-disaster-recovery.md

```diff
@@ -0,0 +1,189 @@
+
+This RFC introduces the first step toward enabling Databend to recover from disasters, such as metadata or data loss, by providing a robust backup and restore solution. Our proposed product, tentatively named `bendsave`, will allow users to back up and restore both metadata and data efficiently.
+
+*The name of this product is not decided yet, let's call it `bendsave`*
+
+### 1. Backup
+
+Create backups of cluster data and metadata using the `bendsave backup` command. Incremental backups are supported, ensuring that only changes since the last backup are saved. This simplifies daily backups.
+
+Example:
+
+```shell
+bendsave backup --from /path/to/query-node-1.toml --to s3://backup/
+```
+
+Key Points:
+- Metadata and data are stored in the backup location.
+- Enables complete cluster recovery, even in cases of total failure.
+
+### 2. List Backups
+
+To view all backups stored in a specified location, use the `bendsave list` command.
+
+Example:
+
+```shell
+bendsave list s3://backup/
+```
+
+### 3. Restore
+
+Restore a Databend cluster from a backup using the `bendsave restore` command. By default, this operates in dry-run mode to prevent accidental restoration. For automatic restoration, use the `--confirm` flag.
+
+Example:
+
+```shell
+# Dry-run mode (default)
+bendsave restore --from s3://backup/path/to/backup/manifest --to /path/to/query-node-1.toml
+
+# Perform the restoration immediately
+bendsave restore --from s3://backup/path/to/backup/manifest --to /path/to/query-node-1.toml --confirm
+```
+
+### 4. Vacuum
+
+Manage backup retention using the `bendsave vacuum` command. This ensures backups adhere to your retention policies by removing old or unnecessary backups.
+
+Example:
+
+```shell
+bendsave vacuum s3://backup \
+    --retention-days 30     \
+    --min-retention-days 7  \
+    --max-backups 5         \
+    --min-backups 2
+```
+
+The `bendsave` tool will provide a simple yet powerful way to secure Databend clusters through backup and restore operations. With features like incremental backups, dry-run restore mode, and vacuum-based retention management, it offers users control and reliability in disaster recovery scenarios.
+
+## Reference-level explanation
+
+`bendsave` will introduce an `BackupManifest` in which stores the following things:
+
+- metadata of given backup: like backup time, backup location, backup type (full or incremental), etc.
+- the locations of metadata backup: the locations which points to the metadata backup.
+- the locations of data backup: the locations which contains all table data.
+
+```rust
+struct BackupManifest {
+    meta: BackupMeta,
+    backed_up_meta: BackedUpMeta,
+    backed_up_data: Vec<BackedUpData>,
+    ...
+}
+
+struct BackupMeta {
+    backup_time: DateTime<Utc>,
+    backup_type: BackupType,
+    ...
+}
+
+struct BackedUpMeta {
+    location: String
+    ...
+}
+
+struct BackedUpData {
+    source_location: String,
+
+    backup_location: String,
+    etag: String,
+}
+```
```

Comment:
The naming of BackupMeta and BackedUpMeta can be confusing and might benefit from clearer alternatives. Here’s a potential improvement to make the structure more intuitive:

```
struct BackupManifest {
    meta: BackupMeta,
    storage_meta: StorageMeta,
    files: Vec<BackupFile>,
    ...
}

struct BackupMeta {
    timestamp: DateTime<Utc>,
    backup_type: BackupType,
    ...
}

struct StorageMeta {
    root_path: String,
    ...
}

enum Checksum {
    ETag(String),
}

struct BackupFile {
    source_path: String,
    backup_path: String,
    checksum: Checksum,
}
```